### ISSUE TITLE
F1-score always 0

### ISSUE THREAD

**Issue by netsesame2** [kaushaltrivedi/fast-bert#28]
I use metrics as [{'name': 'F1-score', 'function': F1}], run the samples data for 4 epoch.

However, after each epoch, I got the F1 score is 0, what's wrong?

```
from fast_bert.learner import *
from fast_bert.metrics import *
from pytorch_pretrained_bert.tokenization import BertTokenizer

from bert_data import *

import torch
from fastai.text import *
import datetime

run_start_time = datetime.datetime.today().strftime('%Y-%m-%d_%H-%M-%S')

LOG_PATH=Path('logs/')  
MODEL_PATH=Path('models/') 

if not LOG_PATH.exists():
  LOG_PATH.mkdir()
import logging

args = {
    "run_text": "my_test",
    "max_seq_length": 512,
    "do_lower_case": True,
    "train_batch_size": 16,
    "learning_rate": 6e-5,
    "num_train_epochs": 12.0,
    "warmup_proportion": 0.002,
    "local_rank": -1,
    "gradient_accumulation_steps": 1,
    "fp16": True,
    "loss_scale": 128
}

logfile = str(LOG_PATH/'log-{}-{}.txt'.format(run_start_time, args["run_text"]))

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(name)s -   %(message)s',
    datefmt='%m/%d/%Y %H:%M:%S',
    handlers=[
        logging.FileHandler(logfile),
        logging.StreamHandler(sys.stdout)
    ])

logger = logging.getLogger()

device = torch.device('cuda')

if torch.cuda.device_count() > 1:
    multi_gpu = True
else:
    multi_gpu = False
    
print('multi_gpu={}'.format('True' if multi_gpu else 'False'))

DATA_PATH = Path('data/sample/data/')     
LABEL_PATH = Path('data/sample/labels')  

BERT_PRETRAINED_MODEL = "bert/bert-base-uncased"

args["do_lower_case"] = True
args["train_batch_size"] = 16
args["learning_rate"] = 6e-5
args["max_seq_length"] = 512
args["fp16"] = True

tokenizer = BertTokenizer.from_pretrained(BERT_PRETRAINED_MODEL, 
                                          do_lower_case=args['do_lower_case'])

label_cols = ["toxic", "severe_toxic", "obscene", "threat", "insult", "identity_hate"]
databunch = BertDataBunch(DATA_PATH, LABEL_PATH, tokenizer, train_file='train.csv', val_file='valid.csv',
                          test_data='test.csv', label_file="labels.csv",
                          text_col="comment_text", label_col=label_cols,
                          bs=args['train_batch_size'], maxlen=args['max_seq_length'], 
                          multi_gpu=multi_gpu, multi_label=True)

#metrics = [{'name': 'accuracy', 'function': accuracy_multilabel}]                          
#metrics = [{'name': 'roc_auc', 'function': roc_auc}]                          
metrics = [{'name': 'F1-score', 'function': F1}]                          
learner = BertLearner.from_pretrained_model(databunch, BERT_PRETRAINED_MODEL, metrics, device, logger, 
                                            is_fp16=args['fp16'], loss_scale=args['loss_scale'], 
                                            multi_gpu=multi_gpu,  multi_label=True)
learner.fit(4, lr=args['learning_rate'], schedule_type="warmup_linear") 

```

**Comment by netsesame2** [comment]
```
06/27/2019 20:23:59 - INFO - pytorch_pretrained_bert.modeling -   Weights of BertForMultiLabelSequenceClassification not initialized from pretrained model: ['classifier.weight', 'classifier.bias']
06/27/2019 20:23:59 - INFO - pytorch_pretrained_bert.modeling -   Weights from pretrained model not used in BertForMultiLabelSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias']
06/27/2019 20:29:34 - INFO - root -   Loss after epoch 0 - 0.17132626342773438
06/27/2019 20:29:34 - INFO - root -   Running evaluation
06/27/2019 20:30:38 - INFO - root -   Eval results:
06/27/2019 20:30:38 - INFO - root -     eval_loss = 0.15239122178819445
06/27/2019 20:30:38 - INFO - root -     metrics = {'F1-score': 0.0}
06/27/2019 20:30:38 - INFO - root -   --------------------------------------------------------------------------------
06/27/2019 20:36:03 - INFO - root -   Loss after epoch 1 - 0.14778070068359375
06/27/2019 20:36:03 - INFO - root -   Running evaluation
06/27/2019 20:37:06 - INFO - root -   Eval results:
06/27/2019 20:37:06 - INFO - root -     eval_loss = 0.15134538922991073
06/27/2019 20:37:06 - INFO - root -     metrics = {'F1-score': 0.0008571429061703384}
06/27/2019 20:37:06 - INFO - root -   --------------------------------------------------------------------------------
06/27/2019 20:42:31 - INFO - root -   Loss after epoch 2 - 0.14610467529296875
06/27/2019 20:42:31 - INFO - root -   Running evaluation
06/27/2019 20:43:34 - INFO - root -   Eval results:
06/27/2019 20:43:34 - INFO - root -     eval_loss = 0.1521267361111111
06/27/2019 20:43:34 - INFO - root -     metrics = {'F1-score': 0.0}
06/27/2019 20:43:34 - INFO - root -   --------------------------------------------------------------------------------
06/27/2019 20:49:00 - INFO - root -   Loss after epoch 3 - 0.146149658203125
06/27/2019 20:49:00 - INFO - root -   Running evaluation
06/27/2019 20:50:03 - INFO - root -   Eval results:
06/27/2019 20:50:03 - INFO - root -     eval_loss = 0.15223766508556547
06/27/2019 20:50:03 - INFO - root -     metrics = {'F1-score': 0.0}
06/27/2019 20:50:03 - INFO - root -   --------------------------------------------------------------------------------
```

**Comment by Pawel-Kranzberg** [comment]
@netsesame2 - F1 (and F-score in general, as well as simple accuracy, precision and recall) does not work with multi-label classification. Use Hamming loss, EMR or ROC instead. 

BTW, I plan to add further multi-label metrics.

**Comment by netsesame2** [comment]
@Pawel-Kranzberg Thanks for reply. 

I inspect the whole train data, which are so sparse data, many zero values. As a result, the accuracy can reach 99.9% each epoch, however, AUC/ROC value will only 76%.

Looking forward to your new metrics.

**Comment by netsesame2** [comment]
BTW: I got some info during training, it seems that Bert's pretained model can't be used. What's wrong?

> 06/27/2019 20:23:59 - INFO - pytorch_pretrained_bert.modeling -   Weights of BertForMultiLabelSequenceClassification not initialized from pretrained model: ['classifier.weight', 'classifier.bias']
06/27/2019 20:23:59 - INFO - pytorch_pretrained_bert.modeling -   Weights from pretrained model not used in BertForMultiLabelSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias']
```

**Comment by wojtekcz** [comment]
> BTW: I got some info during training, it seems that Bert's pretained model can't be used. What's wrong?
> 
> > 06/27/2019 20:23:59 - INFO - pytorch_pretrained_bert.modeling -   Weights of BertForMultiLabelSequenceClassification not initialized from pretrained model: ['classifier.weight', 'classifier.bias']
> > 06/27/2019 20:23:59 - INFO - pytorch_pretrained_bert.modeling -   Weights from pretrained model not used in BertForMultiLabelSequenceClassification: ['cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias']

This is normal for Bert classifier to not load these weights.
